README.md
# TreeHugger (Sortable Menus)

[![Build Status](https://travis-ci.org/symbiote/silverstripe-treehugger.svg?branch=master)](https://travis-ci.org/symbiote/silverstripe-treehugger)
[![Latest Stable Version](https://poser.pugx.org/symbiote/silverstripe-treehugger/version.svg)](https://github.com/symbiote/silverstripe-treehugger/releases)
[![Latest Unstable Version](https://poser.pugx.org/symbiote/silverstripe-treehugger/v/unstable.svg)](https://packagist.org/packages/symbiote/silverstripe-treehugger)
[![Total Downloads](https://poser.pugx.org/symbiote/silverstripe-treehugger/downloads.svg)](https://packagist.org/packages/symbiote/silverstripe-treehugger)
[![License](https://poser.pugx.org/symbiote/silverstripe-treehugger/license.svg)](https://github.com/symbiote/silverstripe-treehugger/blob/master/LICENSE.md)

Add additional menus (such as footer, sidebars) programmatically that can be managed by CMS users nicely on the SiteConfig or Multisite view.

## Composer Install

```
composer require symbiote/silverstripe-treehugger:~3.0
```

## Features

- Page CMS editing, Adds a checkbox to the "Settings" tab for each defined menu kind.
- Allow re-ordering of pages for each menu kind (independent of the ordering in the site tree)
    - Requires [GridField Extensions](https://github.com/symbiote/silverstripe-gridfieldextensions).
- Easy to drop-in and use with partial caching.
    - Partial caching is recommended to improve page-load times when you have thousands of pages.

## Requirements

* SilverStripe 4.0+
* (Optional) SiteConfig
* (Optional) [Multisites](https://github.com/symbiote/silverstripe-multisites)
* (Optional) [GridField Extensions 2.X or 3.X](https://github.com/symbiote/silverstripe-gridfieldextensions)

**NOTE: To manage menus at the top-level of the site, you must have either SiteConfig or Multisites installed. If you want the ability to sort the menu, you need to install GridField Extensions.**

## Documentation

* [Quick Start](docs/en/quick-start.md)
* [License](LICENSE.md)
* [Contributing](CONTRIBUTING.md)

## Credits

* [Jake Bentvelzen](https://github.com/SilbinaryWolf) for the initial build


CHANGELOG.md
# Changelog

For more information on changes and releases, please visit [Releases](https://github.com/symbiote/silverstripe-treehugger/releases).

This project adheres to [Semantic Versioning](http://semver.org/).


tests/SortablePage.php
<?php

class SortablePage extends Page
{
    private static $extensions = array(
        'SortableMenu',
    );
}


tests/SortableMenuTest.php
<?php

class SortableMenuTest extends SapphireTest
{
    protected $usesDatabase = true;

    public function setUp()
    {
        Config::inst()->update('SortableMenu', 'menus', array(
            'ShowInFooter' => array(
                'Title' => 'Footer',
            ),
            'ShowInSidebar' => array(
                'Title' => 'Sidebar',
            ),
        ));
        parent::setUp();
    }

    public function testDBFieldsApplyToDataObject()
    {
        $record = new SortablePage();
        $record->Title = 'Footer Menu Item #1';
        $record->ShowInFooter = true;
        $record->write();

        // NOTE(Jake): 2018-08-09
        //
        // We re-retrieve the record. If it saved correctly
        // into the database, ShowInFooter will be `true`
        //
        $record = SortablePage::get()->byID($record->ID);
        $this->assertEquals(
            true,
            $record->ShowInFooter,
            'ShowInFooter is not "true". This probably means that the \'menus\' Config/YML in the setUp() method isn\'t working as expected.'
        );

        $record = new SortablePage();
        $record->Title = 'Sidebar Menu Item #1';
        $record->ShowInSidebar = true;
        $record->write();

        // NOTE(Jake): 2018-08-09
        //
        // We re-retrieve the record. If it saved correctly
        // into the database, ShowInSidebar will be `true`
        //
        $record = SortablePage::get()->byID($record->ID);
        $this->assertEquals(
            true,
            $record->ShowInSidebar,
            'ShowInSidebar is not "true". This probably means that the \'menus\' Config/YML in the setUp() method isn\'t working as expected.'
        );
    }

    public function testRetrievingPagesByMenu()
    {
        // NOTE(Jake): 2018-08-09
        //
        // To ensure PostgreSQL support works SS 3.2 projects,
        // I need to pass in 0 or 1, not false or true.
        //
        // Otherwise I get the following error:
        // - pg_query_params(): Query failed: ERROR:  invalid input syntax for integer: ""
        //
        // I assume this is because PostgreSQL just casts the given value to a string and
        // in PHP when you cast `false` to a string, you get a blank string.
        //

        // Check we have no items in the menu
        $count = SortablePage::get()->filter(array('ShowInFooter' => 0))->count();
        $this->assertEquals(0, $count);
        $count = SortablePage::get()->filter(array('ShowInSidebar' => 0))->count();
        $this->assertEquals(0, $count);

        // Check we have 1 footer menu item
        $record = new SortablePage();
        $record->Title = 'Footer Menu Item #1';
        $record->ShowInFooter = true;
        $record->write();
        $count = SortablePage::get()->filter(array('ShowInFooter' => 1))->count();
        $this->assertEquals(1, $count);

        // Check we have 1 sidemenu menu item
        $record = new SortablePage();
        $record->Title = 'Footer Menu Item #1';
        $record->ShowInSidebar = true;
        $record->write();
        $count = SortablePage::get()->filter(array('ShowInSidebar' => 1))->count();
        $this->assertEquals(1, $count);
    }
}


tests/SortableMenuDBTest.php
<?php

namespace Symbiote\SortableMenu\Tests;

use Page;
use SilverStripe\Core\Config\Config;
use Symbiote\SortableMenu\SortableMenuExtension;
use SilverStripe\Dev\FunctionalTest;

class SortableMenuDBTest extends SortableMenuTest
{
    protected static $use_draft_site = true;

    protected $requireDefaultRecordsFrom = [
        Page::class,
    ];

    protected $usesDatabase = true;

    public static function setUpBeforeClass()
    {
        parent::setUpBeforeClass();
        // NOTE(Jake): 2018-08-13
        //
        // Add configs to SortableMenuExtension, then apply to
        // `Page` record. Because this modifies the DB fields, we need
        // to call `static::resetDBSchema(true, true);`
        //
        Config::modify()->set(SortableMenuExtension::class, 'menus', array(
            'ShowInFooter' => array(
                'Title' => 'Footer',
            ),
            'ShowInSidebar' => array(
                'Title' => 'Sidebar',
            ),
        ));
        Page::add_extension(SortableMenuExtension::class);
        static::resetDBSchema(true, true);
    }

    public function testDBFieldsApplyToDataObject()
    {
        $record = new Page();
        $record->Title = 'Footer Menu Item #1';
        $record->ShowInFooter = true;
        $record->write();

        // NOTE(Jake): 2018-08-09
        //
        // We re-retrieve the record. If it saved correctly
        // into the database, ShowInFooter will be `true`
        //
        $record = Page::get()->byID($record->ID);
        $this->assertEquals(
            true,
            $record->ShowInFooter,
            'ShowInFooter is not "true". This probably means that the \'menus\' Config/YML in the setUp() method isn\'t working as expected.'
        );

        $record = new Page();
        $record->Title = 'Sidebar Menu Item #1';
        $record->ShowInSidebar = true;
        $record->write();

        // NOTE(Jake): 2018-08-09
        //
        // We re-retrieve the record. If it saved correctly
        // into the database, ShowInSidebar will be `true`
        //
        $record = Page::get()->byID($record->ID);
        $this->assertEquals(
            true,
            $record->ShowInSidebar,
            'ShowInSidebar is not "true". This probably means that the \'menus\' Config/YML in the setUp() method isn\'t working as expected.'
        );
    }

    public function testRetrievingPagesByMenuDirectly()
    {
        // NOTE(Jake): 2018-08-09
        //
        // To ensure PostgreSQL support works SS 3.2 projects,
        // I need to pass in 0 or 1, not false or true.
        //
        // Otherwise I get the following error:
        // - pg_query_params(): Query failed: ERROR:  invalid input syntax for integer: ""
        //
        // I assume this is because PostgreSQL just casts the given value to a string and
        // in PHP when you cast `false` to a string, you get a blank string.
        //

        // Check we have no items in the menu
        $count = Page::get()->filter(array('ShowInFooter' => 0))->count();
        $this->assertEquals(0, $count);
        $count = Page::get()->filter(array('ShowInSidebar' => 0))->count();
        $this->assertEquals(0, $count);

        // Check we have 1 footer menu item
        $record = new Page();
        $record->Title = 'Footer Menu Item #1';
        $record->ShowInFooter = true;
        $record->write();
        $count = Page::get()->filter(array('ShowInFooter' => 1))->count();
        $this->assertEquals(1, $count);

        // Check we have 1 sidemenu menu item
        $record = new Page();
        $record->Title = 'Footer Menu Item #1';
        $record->ShowInSidebar = true;
        $record->write();
        $count = Page::get()->filter(array('ShowInSidebar' => 1))->count();
        $this->assertEquals(1, $count);
    }
}


code/SortableMenu.php
<?php

class SortableMenu extends DataExtension
{
    private static $menus = array();

    protected static $_cached_cache_max_lastedited = array();

    protected static $inConfigCall = false;

    public function __construct()
    {
        parent::__construct();
    }

    public static function get_extra_config($class, $extension, $args)
    {
        if (self::$inConfigCall) {
            throw new SortableMenuException(__FUNCTION__.': Recursion error.');
        }
        // NOTE(Jake): If 'get_extra_config' is really aiming to be deprecated
        //             post 3.2+, then this logic can moved to a CompositeDBField
        //             and the setup method can migrate to that.
        $menus = static::get_sortable_menu_configuration();
        $dbFields = array();
        foreach ($menus as $fieldName => $extraInfo) {
            $dbFields[$fieldName] = 'Boolean';
            $dbFields[$extraInfo['Sort']] = 'Int';
        }
        $result = array(
            'db' => $dbFields,
        );
        return $result;
    }

    /**
     * @return DataList
     */
    public function SortableMenu($fieldName)
    {
        return $this->SortableMenuUncached($fieldName);
    }

    /**
     * @return string
     */
    public function SortableMenuCacheKey($cacheKey)
    {
        if (!$cacheKey) {
            throw new SortableMenuException(__FUNCTION__.': Cannot have empty $cacheKey parameter.');
        }
        $baseClass = $this->ownerBaseClass;
        if (!isset(self::$_cached_cache_max_lastedited[$baseClass])) {
            // Reuse the max('LastEdited') / count() values for the class if
            // already queried. (Can shave off upto ~0.0010s per sortable menu)
            $list = DataList::create($baseClass);
            self::$_cached_cache_max_lastedited[$baseClass] = implode('_', array(
                $list->max('LastEdited'),
                (int)$list->count(),
            ));
        }
        return $cacheKey.'_'.self::$_cached_cache_max_lastedited[$baseClass];
    }

    /**
     * @return DataList
     */
    public function SortableMenuUncached($fieldName)
    {
        $menus = $this->getSortableMenuConfiguration();
        if (!isset($menus[$fieldName])) {
            throw new SortableMenuException(__FUNCTION__.': "'.$fieldName.'" hasn\'t been configured.');
        }
        $extraInfo = $menus[$fieldName];
        $class = $this->ownerBaseClass;
        $list = DataList::create($class)->filter(array(
            $fieldName => 1
        ))->sort($extraInfo['Sort']);
        // Store fieldname directly on list to use as a cache parameter
        $list->SortableMenuFieldName = $fieldName;
        return $list;
    }

    public function updateSettingsFields($fields)
    {
        $menus = Config::inst()->get(__CLASS__, 'menus');
        if (!$menus) {
            $menus = array();
        }
        foreach (Config::inst()->get(__CLASS__, 'menus') as $fieldName => $extraInfo) {
            $fieldTitle = $fieldName;
            if (isset($extraInfo['Title'])) {
                $fieldTitle = 'Show in "'.$extraInfo['Title'].'"?';
            }
            $fields->insertAfter(CheckboxField::create($fieldName, $fieldTitle), 'ShowInMenus');
        }
    }

    public function getSortableMenuConfiguration()
    {
        return static::get_sortable_menu_configuration();
    }

    public static function get_sortable_menu_configuration()
    {
        self::$inConfigCall = true;
        $menus = Config::inst()->get(__CLASS__, 'menus');
        self::$inConfigCall = false;
        if (!$menus) {
            $menus = array();
        }
        $result = array();
        foreach ($menus as $fieldName => $extraInfo) {
            if (!isset($extraInfo['Sort'])) {
                $extraInfo['Sort'] = 'Sort'.$fieldName;
            }
            if (!isset($extraInfo['Title'])) {
                $extraInfo['Title'] = $fieldName;
            }
            $result[$fieldName] = $extraInfo;
        }
        return $result;
    }
}


src/SortableMenuExtension.php
<?php

namespace Symbiote\SortableMenu;

use Page;
use Symbiote\SortableMenu\SortableMenuExtensionException;
use SilverStripe\ORM\FieldType\DBBoolean;
use SilverStripe\ORM\DataList;
use SilverStripe\Core\Config\Config;
use SilverStripe\Forms\CheckboxField;
use SilverStripe\ORM\DataExtension;
use SilverStripe\ORM\DataObject;
use SilverStripe\CMS\Model\SiteTree;
use Symbiote\Multisites\Model\Site;

class SortableMenuExtension extends DataExtension
{
    private static $menus = array();

    protected static $_cached_cache_max_lastedited = array();

    protected static $inConfigCall = false;

    public function __construct()
    {
        parent::__construct();
    }

    public static function get_extra_config($class, $extension, $args)
    {
        if (self::$inConfigCall) {
            throw new SortableMenuException(__FUNCTION__.': Recursion error.');
        }
        // NOTE(Jake): If 'get_extra_config' is really aiming to be deprecated
        //             post 3.2+, then this logic can moved to a CompositeDBField
        //             and the setup method can migrate to that.
        $menus = static::get_sortable_menu_configuration();
        $dbFields = array();
        foreach ($menus as $fieldName => $extraInfo) {
            $dbFields[$fieldName] = DBBoolean::class;
            $dbFields[$extraInfo['Sort']] = 'Int';
        }
        $result = array(
            'db' => $dbFields,
        );
        return $result;
    }

    /**
     * @return DataList
     */
    public function SortableMenu($fieldName)
    {
        return $this->SortableMenuUncached($fieldName);
    }

    /**
     * @return string
     */
    public function SortableMenuCacheKey($cacheKey)
    {
        if (!$cacheKey) {
            throw new SortableMenuException(__FUNCTION__.': Cannot have empty $cacheKey parameter.');
        }
        $baseClass = $this->getOwnerBaseclass();
        if (!isset(self::$_cached_cache_max_lastedited[$baseClass])) {
            // Reuse the max('LastEdited') / count() values for the class if
            // already queried. (Can shave off upto ~0.0010s per sortable menu)
            $list = DataList::create($baseClass);
            self::$_cached_cache_max_lastedited[$baseClass] = implode('_', array(
                $list->max('LastEdited'),
                (int)$list->count(),
            ));
        }
        // Add SiteID to cache key so each Site has a seperate cache key
        if (class_exists(Site::class)) {
            $record = $this->owner;
            $siteID = 0;
            if ($record instanceof Site) {
                $siteID = $record->ID;
            } else {
                $siteID = $record->SiteID;
            }
            $cacheKey .= '_'.$siteID;
        }
        return $cacheKey.'_'.self::$_cached_cache_max_lastedited[$baseClass];
    }

    /**
     * @return DataList
     */
    public function SortableMenuUncached($fieldName)
    {
        $menus = $this->getSortableMenuConfiguration();
        if (!isset($menus[$fieldName])) {
            throw new SortableMenuException(__FUNCTION__.': "'.$fieldName.'" hasn\'t been configured.');
        }
        $extraInfo = $menus[$fieldName];
        $class = $this->getOwnerBaseclass();
        $list = DataList::create($class)->filter(array(
            $fieldName => 1
        ))->sort($extraInfo['Sort']);
        // Filter by SiteID
        if (class_exists(Site::class)) {
            $record = $this->owner;
            $siteID = 0;
            if ($record instanceof Site) {
                $siteID = $record->ID;
            } else {
                $siteID = $record->SiteID;
            }
            if ($siteID) {
                $list = $list->filter('SiteID', $siteID);
            }
        }
        // Store fieldname directly on list to use as a cache parameter
        $list->SortableMenuFieldName = $fieldName;
        return $list;
    }

    public function updateSettingsFields($fields)
    {
        $menus = Config::inst()->get(__CLASS__, 'menus');
        if (!$menus) {
            $menus = array();
        }
        foreach (Config::inst()->get(__CLASS__, 'menus') as $fieldName => $extraInfo) {
            $fieldTitle = $fieldName;
            if (isset($extraInfo['Title'])) {
                $fieldTitle = 'Show in "'.$extraInfo['Title'].'"?';
            }
            $fields->insertAfter(CheckboxField::create($fieldName, $fieldTitle), 'ShowInMenus');
        }
    }

    public function getSortableMenuConfiguration()
    {
        return static::get_sortable_menu_configuration();
    }

    public static function get_sortable_menu_configuration()
    {
        self::$inConfigCall = true;
        $menus = Config::inst()->get(__CLASS__, 'menus');
        self::$inConfigCall = false;
        if (!$menus) {
            $menus = array();
        }
        $result = array();
        foreach ($menus as $fieldName => $extraInfo) {
            if (!isset($extraInfo['Sort'])) {
                $extraInfo['Sort'] = 'Sort'.$fieldName;
            }
            if (!isset($extraInfo['Title'])) {
                $extraInfo['Title'] = $fieldName;
            }
            $result[$fieldName] = $extraInfo;
        }
        return $result;
    }

    /**
     * @return string
     */
    private function getOwnerBaseclass()
    {
        $owner = $this->getOwner();
        if ($owner instanceof Page) {
            return Page::class;
        }
        // Fallback to support SiteTree for BC reasons.
        // ie. upgrading a SS 3.X wherein this extension was applied to `SiteTree`
        if ($owner instanceof SiteTree) {
            return SiteTree::class;
        }
        $class = get_class($owner);
        $class = DataObject::getSchema()->baseDataClass($class);
        return $class;
    }
}


code/SortableMenuSiteExtension.php
<?php

class SortableMenuSiteExtension extends Extension {
	public function updateSiteCMSFields(FieldList $fields) {
		$tab = $fields->findOrMakeTab('Root.SortableMenu', 'Menus');
		$menus = singleton('SortableMenu')->getSortableMenuConfiguration();
		foreach ($menus as $fieldName => $extraInfo) {
			$fields->addFieldToTab('Root.SortableMenu', $this->owner->createMenuGridField('SiteTree', $fieldName, $extraInfo['Title'], $extraInfo['Sort']));
		}
	}

	public function createMenuGridField($class, $fieldName, $fieldTitle, $sortFieldName) {
		$record = singleton($class);
		$list = $record->SortableMenu($fieldName);

		$gridField = GridField::create($fieldName, $fieldTitle, $list, $config = GridFieldConfig_RelationEditor::create());
		$gridField->setDescription('Any "Modified" or "Draft" pages must be saved and published after sorting to display on the live site.');
		$tab = new Tab($fieldName, $fieldTitle);
		$tab->push($gridField);

		$config->removeComponentsByType('GridFieldAddExistingAutocompleter');
		$config->removeComponentsByType('GridFieldAddNewButton');
		$config->removeComponentsByType('GridFieldEditButton');
		$config->removeComponentsByType('GridFieldDeleteAction');
		$config->removeComponentsByType('GridFieldFilterHeader');
		$config->addComponent(Injector::inst()->createWithArgs('GridFieldAddSortableMenuItem', array($fieldName)));
		$config->addComponent(GridFieldOrderableRows::create($sortFieldName));
		$dataColumns = $config->getComponentByType('GridFieldDataColumns');
		if ($dataColumns) {
			$dataColumns->setDisplayFields(array(
				'ID' => 'ID', 
				'getTreeTitle' => 'Title',
			));
			$dataColumns->setFieldFormatting(array(
                'getTreeTitle' => function($value, &$page) {
                    return sprintf(
                        '<a class="action-detail" href="%s">%s</a>',
                        Controller::join_links(
                            // CMSPageEditController - for regular fields
                            singleton('CMSPageSettingsController')->Link('show'),
                            (int)$page->ID
                        ),
                        $page->TreeTitle // returns HTML, does its own escaping
                    );
                }
            ));
		}
		return $tab;
	}
}

code/GridFieldAddSortableMenuItem.php
<?php

namespace Symbiote\SortableMenu;

use SilverStripe\Forms\GridField\GridField;
use SilverStripe\ORM\DataList;
use SilverStripe\Forms\GridField\GridFieldAddExistingAutocompleter;

/**
 * Like 'GridFieldAddExistingAutocompleter' but instead updates a boolean value of a record when added.
 */
class GridFieldAddSortableMenuItem extends GridFieldAddExistingAutocompleter
{
    protected $fieldName = '';

    protected $sortFieldName = '';

    public function __construct($fieldName, $sortFieldName = null, $targetFragment = 'before', $searchFields = null)
    {
        parent::__construct($targetFragment, $searchFields);
        $this->fieldName = $fieldName;
        $this->sortFieldName = $sortFieldName;
        if (!$this->sortFieldName) {
            $this->sortFieldName = 'Sort'.$this->fieldName;
        }
    }

    public function handleAction(GridField $gridField, $actionName, $arguments, $data)
    {
        switch ($actionName) {
            case 'addto':
                if (isset($data['relationID']) && $data['relationID']) {
                    $gridField->State->GridFieldAddRelation = $data['relationID'];
                    $class = $gridField->getList()->dataClass();
                    $record = DataList::create($class)->filter(array(
                        'ID' => $data['relationID'],
                    ))->first();
                    if ($record) {
                        if (!$record->{$this->fieldName}) {
                            if (!$record->canEdit()) {
                                return;
                            }
                            $record->{$this->fieldName} = 1;
                            if ($record->{$this->sortFieldName} == 0) {
                                $maxSort = DataList::create($class)->filter(array(
                                    $this->fieldName => 1,
                                ))->max($this->sortFieldName) + 1;
                                $record->{$this->sortFieldName} = $maxSort;
                            }
                            $record->write();
                        }
                    }
                }
                break;
        }
    }
}


src/SortableMenuManageExtension.php
<?php

namespace Symbiote\SortableMenu;

use Page;
use SilverStripe\Forms\FieldList;
use SilverStripe\CMS\Model\SiteTree;
use SilverStripe\Forms\TabSet;
use SilverStripe\Forms\Tab;
use SilverStripe\Forms\GridField\GridFieldConfig_RelationEditor;
use SilverStripe\Forms\GridField\GridField;
use SilverStripe\Forms\GridField\GridFieldAddExistingAutocompleter;
use SilverStripe\Forms\GridField\GridFieldAddNewButton;
use SilverStripe\Forms\GridField\GridFieldEditButton;
use SilverStripe\Forms\GridField\GridFieldDeleteAction;
use SilverStripe\Forms\GridField\GridFieldFilterHeader;
use SilverStripe\Core\Injector\Injector;
use SilverStripe\Forms\GridField\GridFieldDataColumns;
use SilverStripe\CMS\Controllers\CMSPageSettingsController;
use SilverStripe\Control\Controller;
use SilverStripe\Core\Extension;
use SilverStripe\ORM\DataList;
use Symbiote\GridFieldExtensions\GridFieldOrderableRows;
use Symbiote\Multisites\Model\Site;

class SortableMenuManageExtension extends Extension
{
    public function updateCMSFields(FieldList $fields)
    {
        // Base class
        $basePageClass = '';
        if (SiteTree::has_extension(SortableMenuExtension::class)) {
            // NOTE(Jake): 2018-08-09
            //
            // An SS 3.X project applied to the SiteTree so we need
            // this for backwards compatibility.
            //
            $basePageClass = SiteTree::class;
        } elseif (Page::has_extension(SortableMenuExtension::class)) {
            $basePageClass = Page::class;
        }

        // If one of the expected classes has the extension, then generate
        // the fields.
        if ($basePageClass !== '') {
            // Setup fields
            $rootTabSet = $fields->fieldByName('Root');
            $sortableMenuTab = $rootTabSet->fieldByName('SortableMenu');
            if (!$sortableMenuTab) {
                $sortableMenuTab = TabSet::create(SortableMenuExtension::class, 'Menus');
                $rootTabSet->push($sortableMenuTab);
            }
            $sortableMenuTab->setTitle('Menus');
            if (!$sortableMenuTab instanceof TabSet) {
                throw new SortableMenuException('Sortable Menu must be a "TabSet", not "'.get_class($sortableMenuTab).'"');
            }
            $sortableMenuTab = $fields->findOrMakeTab('Root.SortableMenu', 'Menus');
            $menus = singleton(SortableMenuExtension::class)->getSortableMenuConfiguration();
            foreach ($menus as $fieldName => $extraInfo) {
                $fieldTitle = $extraInfo['Title'];
                // NOTE(Jake): Check if Tab has already been created by user-code / outside of the module
                $menuTab = $sortableMenuTab->fieldByName($fieldName);
                if (!($menuTab instanceof Tab)) {
                    $menuTab = Tab::create($fieldName, $fieldTitle);
                    $sortableMenuTab->push($menuTab);
                }
                $menuTab->setTitle($fieldTitle);
                $menuTab->push($this->owner->createMenuGridField($basePageClass, $fieldName, $fieldTitle, $extraInfo['Sort']));
            }
        }
    }

    /**
     * Support 'Site' from Multisites module
     */
    public function updateSiteCMSFields(FieldList $fields)
    {
        $this->updateCMSFields($fields);
    }

    public function createMenuGridField($class, $fieldName, $fieldTitle, $sortFieldName)
    {
        if (!class_exists($class)) {
            throw new SortableMenuException($class.' does not exist.');
        }
        if (!$class::has_extension(SortableMenuExtension::class)) {
            throw new SortableMenuException($class.' does not have SortableMenu extension applied.');
        }

        //
        $record = singleton($class);
        $list = $record->SortableMenu($fieldName);
        if ($this->owner instanceof Site) {
            // NOTE(Jake): 2018-08-13
            //
            // Only show pages from this SiteID.
            //
            $siteID = $this->owner->ID;
            if ($siteID) {
                $list = $list->filter('SiteID', $siteID);
            }
        }

        $gridField = GridField::create($fieldName, $fieldTitle, $list, $config = GridFieldConfig_RelationEditor::create());
        $gridField->setDescription('Any "Modified" or "Draft" pages must be saved and published after sorting to display on the live site.');

        $config->removeComponentsByType(GridFieldAddExistingAutocompleter::class);
        $config->removeComponentsByType(GridFieldAddNewButton::class);
        $config->removeComponentsByType(GridFieldEditButton::class);
        $config->removeComponentsByType(GridFieldDeleteAction::class);
        $config->removeComponentsByType(GridFieldFilterHeader::class);
        $gridFieldAddSortableMenu = Injector::inst()->createWithArgs(GridFieldAddSortableMenuItem::class, array($fieldName));
        if ($this->owner instanceof Site) {
            // NOTE(Jake): 2018-08-13
            //
            // Limit pages you can add to the menu by SiteID.
            // If you need to link across websites, please use RedirectorPage.
            //
            $dataClass = $gridField->getModelClass();
            $searchList = DataList::create($dataClass);
            $siteID = $this->owner->ID;
            if ($siteID) {
                $searchList = $searchList->filter('SiteID', $siteID);
            }
            $gridFieldAddSortableMenu->setSearchList($searchList);
        }
        $config->addComponent($gridFieldAddSortableMenu);
        if (class_exists(GridFieldOrderableRows::class)) {
            $config->addComponent(GridFieldOrderableRows::create($sortFieldName));
        }
        $dataColumns = $config->getComponentByType(GridFieldDataColumns::class);
        if ($dataColumns) {
            $dataColumns->setDisplayFields(array(
                'ID' => 'ID',
                'getTreeTitle' => 'Title',
            ));
            $dataColumns->setFieldFormatting(array(
                'getTreeTitle' => function ($value, &$page) {
                    return sprintf(
                        '<a class="action-detail" href="%s">%s</a>',
                        Controller::join_links(
                            // CMSPageEditController - for regular fields
                            singleton(CMSPageSettingsController::class)->Link('show'),
                            (int)$page->ID
                        ),
                        $page->TreeTitle // returns HTML, does its own escaping
                    );
                }
            ));
        }
        return $gridField;
    }
}
